design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,224 @@
+  how to leverage future Connect improvements
+
+## Value/Return
+
+This proposal targets _usability_ improvements for KSQL and simplifies bootstrapping KSQL for users
+that may already have some, but not all, of their data in Kafka. As with the motivating example,
+KSQL users will be able to easily model external systems within KSQL. The target user has already
+played around with KSQL and is working on prototyping their first real application with existing
+data.
+
+## Design
+
+### Architecture
+
+There are three ways to integrate KSQL with Connect:
+1. Run them separately and supply syntax to communicate between them
+1. Embed Connect into KSQL, managing individual connect workers as part of the KSQL ecosystem
+1. Package Connect and KSQL into a single JVM, but just spin them up at the top level
+
+There is some consensus that at scale KSQL and Connect should run as separate components. This
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**
+
+### Syntax
+
+Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
+actions that simply declare metadata on top of existing data. This solves two issues:
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+There are a few more minor points of consideration:
+- All "DML" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- Some connectors will have KSQL-specific wrapper implementations that will allow simpler 
+  configuration and tighter integration. These connectors will also automatically issue the DDL 
+  statements for any topics that they create, when possible.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
```

Comment:
Is the `connector.class` a required option for the `CREATE EXTERNAL SOURCE`? Also, I see that `DESCRIBE source_name` and `SHOW EXTERNAL SOURCES` has an output with headers that use `connector`. Also we have `GET /connectors`.

Are we planning of using the `CREATE EXTERNAL SOURCE` for other external sources other than connect at some point? Seems we are making this DDL command for Connect only, why not calling it `CREATE CONNECTOR`? Isn't more appropriate for it?

Reply:
Sorry I'm not sure I get all of the questions you have there but I'll try to answer my best:

- `GET /connectors` is the call we'll be making _to_ connect
- For now and for the MVP only connect will be supported
- `CONNECTOR` doesn't differentiate between SOURCE/SINK, so between `CREATE SOURCE CONNECTOR` and `CREATE EXTERNAL SOURCE` I don't think there's much of a difference

Reply:
I think `CREATE SOURCE CONNECTOR` is preferable to `CREATE EXTERNAL SOURCE`.

1. 'source connector' is used in the Connect ecosystem and docs, so easier for people to relate prior experience or reading the docs to KSQL terminology
2. 'source' is currently used in KSQL to refer to a source _table_ or _stream_.  In this sense one might expect`EXTERNAL SOURCE` to be something very similar, but it isn't.
3. using `CREATE SOURCE CONNECTOR` leaves us free to use `CREATE EXTERNAL SOURCE` later something else, e.g. if we enhance Connect/KSQL so that we can ingest single tables and internally multiplex this into a single connector.